File: container-with-most-water.py
from typing import List


class Solution:
    def maxArea(self, height: List[int]) -> int:
        max_area = 0
        lt, rt = 0, len(height) - 1

        while lt < rt:
            lh, rh = height[lt], height[rt]
            area = (rt - lt) * min(lh, rh)
            max_area = max(max_area, area)

            if lh <= rh:
                lt += 1
            else:
                rt -= 1

        return max_area

File: container-with-most-water.cpp
#include <vector>

using namespace std;

class Solution {
public:
  int maxArea(vector<int> &height) {
    int max_area = 0;
    int lt = 0, rt = height.size() - 1;

    while (lt < rt) {
      int lh = height[lt];
      int rh = height[rt];
      max_area = max(max_area, (rt - lt) * min(lh, rh));

      if (lh <= rh)
        ++lt;
      else
        --rt;
    }

    return max_area;
  }
};
